### PR TITLE
Improve performance on .NetFx

### LINKED ISSUE #546: TorchSharp is slower on NetFX comparing to Net Core
We have enabled TorchSharp to run on .NET Framework. We noticed a performance degradation comparing when running against Net Core.  One obvious place to look at is the pinvoke calls cost. measuring simple pinvoke seeing the numbers:

```
|      Method      |     Mean      |    Error   |   StdDev   |  
| NetCore5OPInvoke |      4.371 ns |  0.0107 ns |  0.0095 ns |  
| NetFxPInvoke     |      7.885 ns |  0.0333 ns |  0.0295 ns | 
``` 

Need to look in general on the full stack to find out if there is any opportunity for optimization.

### PR DESCRIPTION
**Pull request by alinpahontu2912**
Fixes #546

Replace [MarshalAs(UnmanagedType.U1)] bool parameters and return values with plain byte in hot-path P/Invoke declarations to eliminate marshalling overhead on .NET Framework.

Non-blittable bool types require the runtime to generate marshalling stubs for each P/Invoke call. On .NET Framework this adds ns per call vs .NET Core. By using byte (a blittable type), the runtime can make direct native calls without marshalling, closing the performance gap.

Converted P/Invoke areas:
- Autograd: grad/backward (retain_graph, create_graph, allow_unused)
- Tensor: pooling, device transfer, gradient, reductions, sort/search
- NN: dropout, batch_norm, instance_norm, attention, loss functions
- Linear algebra: all bool params (keepdim, hermitian, check_errors)
- CUDA backend: all bool getters/setters
- JIT: module train/is_training
- Data: loader/iterator

Call sites updated to convert bool↔byte at the managed boundary:
- Parameters: (byte)(boolValue ? 1 : 0)
- Returns: nativeResult != 0

### CODE REVIEW

File: test/PInvokeBenchmark/Program.cs [hunk 1]
```diff
@@ -0,0 +1,110 @@
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+namespace PInvokeBenchmark
```
Comment (cincuranet) [inline test/PInvokeBenchmark/Program.cs:5]: Any reason why _BenchmarkDotNet_ is not used here?

Maybe a more general question. We know this issue exists in .NET Framework and I highly doubt that anything is going to change there. Do we need the benchmark at all? I would post the numbers into the PR and possibly the code for benchmark. And that's it. If somebody wants to reevaluate the number (s)he can rerun it from PR and compare.

File: test/PInvokeBenchmark/PInvokeBenchmark.csproj [hunk 2]
```diff
@@ -0,0 +1,12 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFrameworks>net6.0;net472</TargetFrameworks>
+    <LangVersion>9.0</LangVersion>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <StartupObject>PInvokeBenchmark.Program</StartupObject>
```
Comment (cincuranet) [inline test/PInvokeBenchmark/PInvokeBenchmark.csproj:7]: I believe these two lines are not needed.

**Review by cincuranet — COMMENTED** [review]
All these `x ? 1 : 0` made me wonder... Why not use `Convert.ToByte` instead?

**Comment by alinpahontu2912** [comment]
Will take a look again. I'd say it's importatnt to figure out if it's worth spending time to fix this. (Not sure how to do that, as I can t see metrics on how much the package is used for each variant) The issue was opened a long ago when support to netstandard2.0 to TorchSharp was added. Users running on NETFX are expected to see some differences and I am not sure if we want to help much there. Tarek mentioned that we should encourage people to switch to .net core as much as possible. There is also this comment explaining the same:  https://github.com/dotnet/TorchSharp/issues/546#issuecomment-1435991796. What do you think @cincuranet ?

**Comment by cincuranet** [comment]
The numbers that Tarek posted show about ~3.5ns overhead, which is ~80% slower. However, this is the raw interop cost only. In practice, the actual operations behind each P/Invoke call take orders of magnitude longer, so the marshalling overhead is noise for most workloads. It would only matter for extremely tight loops making millions of P/Invoke calls with trivial native work.

Bottom line. The real-world benefit is questionable. .NET Framework's P/Invoke gap is known. We don't have strong evidence of significant .NET Framework usage of TorchSharp. The #546 has basically non-existing community traction since 2022.

I would suggest closing this PR. And closing the issue, explaining that if you want best perf, you should use latest .NET.

**Comment by alinpahontu2912** [comment]
Thanks, I will close the pr and the issue then
